File: client/src/main/java/util/GameSimulator.java
package util;

import java.util.List;
import java.util.Random;

import object.Bid;
import object.Player;
import screen.ScreenCache;
import screen.SimulationDialog;

public final class GameSimulator 
{
	private SimulationParms parms = null;
	private int personToStart = 0;
	
	private Bid lastBid = null;
	
	private Player opponentZero = new Player(0, "");
	private Player opponentOne = new Player(1, "");
	private Player opponentTwo = new Player(2, "");
	private Player opponentThree = new Player(3, "");
	
	private int[] playOrder = {2,3,1,0}; //standard play order
	
	/**
	 * Simulation methods
	 */
	public void startNewGame(int number)
	{
		Debug.appendBanner("Game " + number, parms.getLogging());

		initVariablesForSimulationNewGame();
		initVariablesForSimulationNewRound();
		setRandomPersonToStart();
		
		Player player = getPlayer(personToStart);
		processOpponentTurn(player);
	}

	private void startNewRound()
	{
		Debug.appendBanner("New Round", parms.getLogging());
		log("opponentZeroNumberOfCards = " + opponentZero.getNumberOfCards());
		log("opponentOneNumberOfCards = " + opponentOne.getNumberOfCards());
		log("opponentTwoNumberOfCards = " + opponentTwo.getNumberOfCards());
		log("opponentThreeNumberOfCards = " + opponentThree.getNumberOfCards());

		knockOutPlayers();
		boolean gameOver = checkForResult();

		if (!gameOver)
		{
			initVariablesForSimulationNewRound();

			Player playerToStart = getPlayer(personToStart);
			processOpponentTurn(playerToStart);
		}
	}

	private void populateHands(List<String> deck)
	{
		log("Dealing hands...");

		GameUtil.populateHand(opponentZero, deck, parms.getLogging());
		GameUtil.populateHand(opponentOne, deck, parms.getLogging());
		GameUtil.populateHand(opponentTwo, deck, parms.getLogging());
		GameUtil.populateHand(opponentThree, deck, parms.getLogging());
	}

	private void setRandomPersonToStart()
	{
		Random random = new Random();
		if (opponentTwo.isEnabled() 
		  && opponentThree.isEnabled())
		{
			personToStart = random.nextInt(4);
		}
		else if (opponentTwo.isEnabled() 
		  && !opponentThree.isEnabled())
		{
			personToStart = random.nextInt(3);
		}
		else if (!opponentTwo.isEnabled() 
		  && opponentThree.isEnabled())
		{
			personToStart = random.nextInt(3);
			if (personToStart == 2)
			{
				personToStart++;
			}
		}
		else
		{
			personToStart = random.nextInt(2);
		}
		
		//If we're forcing player zero to start, then set this
		if (parms.getForceStart())
		{
			personToStart = 0;
		}
	}

	private void knockOutPlayers()
	{
		log("Knocking out players...");
		
		knockOut(opponentZero);
		knockOut(opponentOne);
		knockOut(opponentTwo);
		knockOut(opponentThree);
	}
	
	private void knockOut(Player player)
	{
		if (player.getNumberOfCards() == 0
		  && player.isEnabled())
		{
			log("Opponent " + player + " disabled");
			player.setEnabled(false);
		}
	}

	private boolean checkForResult()
	{
		int opponentZeroNumberOfCards = opponentZero.getNumberOfCards();
		int opponentOneNumberOfCards = opponentOne.getNumberOfCards();
		int opponentTwoNumberOfCards = opponentTwo.getNumberOfCards();
		int opponentThreeNumberOfCards = opponentThree.getNumberOfCards();
		
		if (opponentZeroNumberOfCards == 0 
		  && opponentOneNumberOfCards == 0 
		  && opponentTwoNumberOfCards == 0)
		{
			recordWin(opponentThree);
			return true;
		}
		else if (opponentZeroNumberOfCards == 0 
		  && opponentOneNumberOfCards == 0 
		  && opponentThreeNumberOfCards == 0)
		{
			recordWin(opponentTwo);
			return true;
		}
		else if (opponentZeroNumberOfCards == 0 
		  && opponentThreeNumberOfCards == 0 
		  && opponentTwoNumberOfCards == 0)
		{
			recordWin(opponentOne);
			return true;
		}
		else if (opponentOneNumberOfCards == 0 
		  && opponentTwoNumberOfCards == 0 
		  && opponentThreeNumberOfCards == 0)
		{
			recordWin(opponentZero);
			return true;
		}
		return false;
	}
	
	private void recordWin(Player winner)
	{
		SimulationDialog dialog = ScreenCache.get(SimulationDialog.class);
		if (winner.getNumberOfCards() == parms.getNumberOfCards())
		{
			dialog.recordWin(winner.getPlayerNumber(), true);
		}
		else
		{
			dialog.recordWin(winner.getPlayerNumber(), false);
		}
	}
	
	private void processOpponentTurn(Player opponent)
	{	
		if (!opponent.isEnabled())
		{
			int nextPlayerNumber = playOrder[opponent.getPlayerNumber()];
			processOpponentTurn(getPlayer(nextPlayerNumber));
			return;
		}
		
		Debug.appendBanner("Opponent " + opponent, parms.getLogging());
		if (lastBid != null)
		{
			ScreenCache.get(SimulationDialog.class).recordOpportunityToChallenge(opponent.getPlayerNumber());
		}
		
		StrategyParms stratParms = getStrategyParms(opponent);
		Bid bid = CpuStrategies.processOpponentTurn(stratParms, opponent);
		if (bid == null)
		{
			//Abort the simulation, something's gone wrong
			throw new SimulationException("Simulation error");
		}
		
		bid.setPlayer(opponent);

		if (bid.isChallenge())
		{
			processChallenge(opponent);
		}
		else
		{
			lastBid = bid;
			int nextPlayerNumber = playOrder[opponent.getPlayerNumber()];
			processOpponentTurn(getPlayer(nextPlayerNumber));
		}
	}

	private void processChallenge(Player challenger)
	{
		log("Challenged");

		SimulationDialog dialog = ScreenCache.get(SimulationDialog.class);
		if (!lastBid.isOverbid(opponentZero.getHand(), opponentOne.getHand(), opponentTwo.getHand(), 
				opponentThree.getHand(), parms.getJokerValue()))
		{
			log("not overbid");
			dialog.recordChallenge(challenger.getPlayerNumber(), false);
			
			challenger.setCardsToSubtract(1);
			challenger.doSubtraction();
			personToStart = challenger.getPlayerNumber();
		}
		else
		{
			log("overbid");
			dialog.recordChallenge(challenger.getPlayerNumber(), true);
			
			Player bidder = lastBid.getPlayer();
			bidder.setCardsToSubtract(1);
			bidder.doSubtraction();
			personToStart = bidder.getPlayerNumber();
		}
		
		startNewRound();
	}
	
	private void initVariablesForSimulationNewGame()
	{
		opponentZero.setEnabled(true);
		opponentOne.setEnabled(true);
		boolean opponentTwoEnabled = parms.getOpponentTwoEnabled();
		opponentTwo.setEnabled(opponentTwoEnabled);
		boolean opponentThreeEnabled = parms.getOpponentThreeEnabled();
		opponentThree.setEnabled(opponentThreeEnabled);
		int opponentTwoCoeff = opponentTwoEnabled? 1:0;
		int opponentThreeCoeff = opponentThreeEnabled? 1:0;
		
		int numberOfCards = parms.getNumberOfCards();
		opponentZero.setNumberOfCards(numberOfCards);
		opponentOne.setNumberOfCards(numberOfCards);
		opponentTwo.setNumberOfCards(opponentTwoCoeff * numberOfCards);
		opponentThree.setNumberOfCards(opponentThreeCoeff * numberOfCards);
		
		opponentZero.setStrategy(parms.getOpponentZeroStrategy());
		opponentOne.setStrategy(parms.getOpponentOneStrategy());
		opponentTwo.setStrategy(parms.getOpponentTwoStrategy());
		opponentThree.setStrategy(parms.getOpponentThreeStrategy());

		setPlayOrder(parms.getRandomiseOrder());
	}
	
	private void setPlayOrder(boolean randomiseOrder)
	{
		if (randomiseOrder)
		{
			Random dice = new Random();
			int orderNo = dice.nextInt(6);

			if (orderNo == 0)
			{
				playOrder[0] = 1;
				playOrder[1] = 2;
				playOrder[2] = 3;
				playOrder[3] = 0;
			}
			else if (orderNo == 1)
			{
				playOrder[0] = 1;
				playOrder[1] = 3;
				playOrder[2] = 0;
				playOrder[3] = 2;
			}
			else if (orderNo == 2)
			{
				playOrder[0] = 2;
				playOrder[1] = 0;
				playOrder[2] = 3;
				playOrder[3] = 1;
			}
			else if (orderNo == 3)
			{
				playOrder[0] = 2;
				playOrder[1] = 3;
				playOrder[2] = 1;
				playOrder[3] = 0;
			}
			else if (orderNo == 4)
			{
				playOrder[0] = 3;
				playOrder[1] = 0;
				playOrder[2] = 1;
				playOrder[3] = 2;
			}
			else if (orderNo == 5)
			{
				playOrder[0] = 3;
				playOrder[1] = 2;
				playOrder[2] = 0;
				playOrder[3] = 1;
			}
		}
		else
		{
			playOrder[0] = 2;
			playOrder[1] = 3;
			playOrder[2] = 1;
			playOrder[3] = 0;
		}
	}

	private void initVariablesForSimulationNewRound()
	{
		lastBid = null;
		
		opponentZero.resetHand();
		opponentOne.resetHand();
		opponentTwo.resetHand();
		opponentThree.resetHand();
		
		List<String> deck = CardsUtil.createAndShuffleDeck(parms.getIncludeJokers(), 
		  parms.getJokerQuantity(), parms.getIncludeMoons(), parms.getIncludeStars(), parms.getNegativeJacks());
				
		populateHands(deck);
	}
	
	private void log(String text)
	{
		Debug.append(text, parms.getLogging());
	}
	
	private StrategyParms getStrategyParms(Player opponent)
	{
		StrategyParms stratParms = parms.getStrategyParms();
		stratParms.setLastBid(lastBid);
		stratParms.setOpponentOneCards(opponentOne.getNumberOfCards());
		stratParms.setOpponentTwoCards(opponentTwo.getNumberOfCards());
		stratParms.setOpponentThreeCards(opponentThree.getNumberOfCards());
		stratParms.setPlayerCards(opponentZero.getNumberOfCards());
		
		//Set the cards revealed
		stratParms.appendCardsOnShowFromOpponent(opponent, opponentZero);
		stratParms.appendCardsOnShowFromOpponent(opponent, opponentOne);
		stratParms.appendCardsOnShowFromOpponent(opponent, opponentTwo);
		stratParms.appendCardsOnShowFromOpponent(opponent, opponentThree);
		
		return stratParms;
	}
	
	private Player getPlayer(int playerNumber)
	{
		if (playerNumber == 0)
		{
			return opponentZero;
		}
		else if (playerNumber == 1)
		{
			return opponentOne;
		}
		else if (playerNumber == 2)
		{
			return opponentTwo;
		}
		
		return opponentThree;
	}
	
	public void setSimulationParms(SimulationParms parms)
	{
		this.parms = parms;
	}
}

File: client/src/main/kotlin/util/GameSimulator.kt
package util

import game.BidAction
import game.ChallengeAction
import `object`.Player
import screen.ScreenCache.get
import screen.SimulationDialog
import utils.CoreGlobals.logger

class GameSimulator(private val params: SimulationParams) {
    private var personToStart = 0

    private var lastBid: BidAction<*>? = null

    val opponentZero: Player = Player(0, "").also { it.name = "0" }
    val opponentOne: Player = Player(1, "").also { it.name = "1" }
    val opponentTwo: Player = Player(2, "").also { it.name = "2" }
    val opponentThree: Player = Player(3, "").also { it.name = "3" }

    var playOrder = listOf(0, 2, 3, 1) // standard play order

    /** Simulation methods */
    fun startNewGame(number: Int) {
        log("*** Game $number ***")

        initVariablesForSimulationNewGame()
        initVariablesForSimulationNewRound()
        personToStart = computeRandomPersonToStart()

        val player: Player = getPlayer(personToStart)
        processOpponentTurn(player)
    }

    private fun startNewRound() {
        log("*** New Round ***")
        log("opponentZeroNumberOfCards = " + opponentZero.numberOfCards)
        log("opponentOneNumberOfCards = " + opponentOne.numberOfCards)
        log("opponentTwoNumberOfCards = " + opponentTwo.numberOfCards)
        log("opponentThreeNumberOfCards = " + opponentThree.numberOfCards)

        knockOutPlayers()
        val gameOver = checkForResult()

        if (!gameOver) {
            initVariablesForSimulationNewRound()

            val playerToStart: Player = getPlayer(personToStart)
            if (playerToStart.isEnabled) {
                processOpponentTurn(playerToStart)
            } else {
                processOpponentTurn(nextPlayer(playerToStart))
            }
        }
    }

    private fun populateHands(deck: List<String>) {
        log("Dealing hands...")

        allPlayers().forEach { GameUtil.populateHand(it, deck, params.enableLogging) }
    }

    private fun computeRandomPersonToStart(): Int {
        if (params.forceStart) {
            return 0
        }

        return allPlayers().filter(Player::isEnabled).random().playerNumber
    }

    private fun knockOutPlayers() {
        log("Knocking out players...")

        allPlayers().forEach(::knockOut)
    }

    private fun knockOut(player: Player) {
        if (player.numberOfCards == 0 && player.isEnabled) {
            log("Opponent $player disabled")
            player.isEnabled = false
        }
    }

    private fun checkForResult(): Boolean {
        val stillAlive = allPlayers().filter { it.numberOfCards > 0 }
        if (stillAlive.size == 1) {
            val winner = stillAlive.first()
            val perfectGame = winner.numberOfCards == params.settings.startingCards
            get(SimulationDialog::class.java).recordWin(winner.playerNumber, perfectGame)
            return true
        }

        return false
    }

    private fun allPlayers() = playOrder.map(::getPlayer)

    private fun allCards() = allPlayers().flatMap { it.hand }

    private fun processOpponentTurn(opponent: Player) {
        if (!opponent.isEnabled) {
            throw Exception("Trying to take turn for $opponent, but they're disabled")
        }

        log("*** Opponent $opponent ***")
        if (lastBid != null) {
            get(SimulationDialog::class.java).recordOpportunityToChallenge(opponent.playerNumber)
        }

        val stratParms = getStrategyParms(opponent)
        val action =
            CpuStrategies.processOpponentTurn(stratParms, opponent)
                ?: // Abort the simulation, something's gone wrong
                throw SimulationException("Simulation error")

        if (action is ChallengeAction) {
            processChallenge(opponent)
        } else {
            lastBid = action as BidAction<*>
            processOpponentTurn(nextPlayer(opponent))
        }
    }

    fun nextPlayer(opponent: Player): Player {
        val currentPlayerIndex = playOrder.indexOf(opponent.playerNumber)
        val nextIndices = (1..3).map { (currentPlayerIndex + it) % playOrder.size }
        val nextPlayers = nextIndices.map { getPlayer(playOrder[it]) }.filter { it.isEnabled }

        if (nextPlayers.isEmpty()) {
            throw Exception(
                "No valid next player found. Play order is $playOrder, currentPlayer is ${opponent.playerNumber}, nextIndices: $nextIndices"
            )
        }

        return nextPlayers.first()
    }

    private fun processChallenge(challenger: Player) {
        log("Challenged")

        val lastBid = lastBid ?: throw Exception("Processing challenge with no lastBid")

        val dialog = get(SimulationDialog::class.java)
        if (!lastBid.isOverbid(allCards(), params.settings)) {
            log("not overbid")
            dialog.recordChallenge(challenger.playerNumber, false)

            challenger.cardsToSubtract = 1
            challenger.doSubtraction()
            personToStart = challenger.playerNumber
        } else {
            log("overbid")
            dialog.recordChallenge(challenger.playerNumber, true)

            val name = lastBid.playerName
            val bidder = getPlayer(name.toInt())
            bidder.cardsToSubtract = 1
            bidder.doSubtraction()
            personToStart = bidder.playerNumber
        }

        startNewRound()
    }

    private fun initVariablesForSimulationNewGame() {
        opponentZero.isEnabled = true
        opponentOne.isEnabled = true
        val opponentTwoEnabled = params.opponentTwoEnabled
        opponentTwo.isEnabled = opponentTwoEnabled
        val opponentThreeEnabled = params.opponentThreeEnabled
        opponentThree.isEnabled = opponentThreeEnabled
        val opponentTwoCoeff = if (opponentTwoEnabled) 1 else 0
        val opponentThreeCoeff = if (opponentThreeEnabled) 1 else 0

        val numberOfCards = params.settings.startingCards
        opponentZero.numberOfCards = numberOfCards
        opponentOne.numberOfCards = numberOfCards
        opponentTwo.numberOfCards = opponentTwoCoeff * numberOfCards
        opponentThree.numberOfCards = opponentThreeCoeff * numberOfCards

        opponentZero.strategy = params.opponentZeroStrategy
        opponentOne.strategy = params.opponentOneStrategy
        opponentTwo.strategy = params.opponentTwoStrategy
        opponentThree.strategy = params.opponentThreeStrategy

        setPlayOrder(params.randomiseOrder)
    }

    private fun setPlayOrder(randomiseOrder: Boolean) {
        playOrder =
            if (randomiseOrder) {
                playOrder.shuffled()
            } else {
                listOf(2, 3, 1, 0)
            }

        log("Player order: $playOrder")
    }

    private fun initVariablesForSimulationNewRound() {
        lastBid = null

        opponentZero.resetHand()
        opponentOne.resetHand()
        opponentTwo.resetHand()
        opponentThree.resetHand()

        val deck = CardsUtil.createAndShuffleDeck(params.settings)

        populateHands(deck)
    }

    private fun log(text: String) {
        if (params.enableLogging) {
            logger.info("simulation", text)
        }
    }

    private fun getStrategyParms(opponent: Player): StrategyParams {
        val totalCards = allPlayers().sumOf(Player::getNumberOfCards)
        val otherPlayers = allPlayers().filterNot { it == opponent }
        val cardsOnShow = otherPlayers.flatMap { it.revealedCards }

        return StrategyParams(
            params.settings,
            totalCards,
            cardsOnShow,
            lastBid,
            params.enableLogging,
        )
    }

    private fun getPlayer(playerNumber: Int) =
        when (playerNumber) {
            0 -> opponentZero
            1 -> opponentOne
            2 -> opponentTwo
            else -> opponentThree
        }
}
